### PR TITLE
Openstack Security group support for UDP and ICMP

### PR DESCRIPTION
**Pull request by gadresushrut**
Also setting the default values for TCP, UDP and ICMP.

### CODE REVIEW

File: app/scripts/modules/openstack/common/validateType.directive.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+])
+.directive('validateType', [
+  function() {
+
+    var link = function($scope, $element, $attrs, ctrl) {
+
+      var validate = function(viewValue) {
+        var comparisonModel = $attrs.validateType;
+
+        if(parseInt(viewValue, 10) === -1 && parseInt(comparisonModel, 10) > -1)
```
Comment (varikin) [inline app/scripts/modules/openstack/common/validateType.directive.js]: What does it mean if `viewValue` is `-1` and `comparisonModel` is greater than `-1`?

Comment (gadresushrut) [inline app/scripts/modules/openstack/common/validateType.directive.js]: Actually, there is a validation on ICMP on devstack.
If viewValue is type for ICMP and comparisonModel is code.
So is viewValue [type] == -1 and ComparisonModel[code] > -1 then give it an error

File: app/scripts/modules/openstack/securityGroup/configure/wizard/rules.controller.js [hunk 2]
```diff
@@ -4,13 +4,107 @@ let angular = require('angular');
...
+           $scope.availableSecurityGroups = securityGroups;
+         }
+
+         //Add CIDR at the start of avaibleSecurityGroup Collection.
+         var cidrObj = {
+           id: 'CIDR',
+           name: 'CIDR'
+         };
+
+         if($scope.availableSecurityGroups.unshift !== undefined) {
```
Comment (anotherchrisberry) [inline app/scripts/modules/openstack/securityGroup/configure/wizard/rules.controller.js]: I think you can simplify this whole block to:
```javascript
if ($scope.availableSecurityGroups.unshift && $scope.availableSecurityGroups.some(g => g.id === 'CIDR')) {
  $scope.availableSecurityGroups.unshift(cidrObj);
}
```

File: app/scripts/modules/openstack/securityGroup/configure/wizard/rules.controller.js [hunk 3]
```diff
@@ -4,13 +4,107 @@ let angular = require('angular');
...
+          }
+         }
+         $scope.allSecurityGroupsUpdated.onNext();
+       });
+    };
+
+    $scope.remoteSecurityGroupSelected = function(indx, remoteSecurityGroupId) {
+      var rule = $scope.securityGroup.rules[indx];
+      if(remoteSecurityGroupId === 'CIDR')
+      {
```
Comment (anotherchrisberry) [inline app/scripts/modules/openstack/securityGroup/configure/wizard/rules.controller.js]: Bracket should be on the line above, i.e.
```javascript
if (remoteSecurityGroupId === 'CIDR') {
```

File: app/scripts/modules/openstack/securityGroup/transformer.js [hunk 4]
```diff
@@ -23,22 +24,43 @@ module.exports = angular.module('spinnaker.openstack.securityGroup.transformer',
...
+          icmpCode : -1
       };
     }
 
-    function prepareForEdit(securityGroup) {
+    function prepareForSaving(securityGroup) {
+      _.forEach(securityGroup.rules, function(value) {
+
+        if(value['remoteSecurityGroupId'] === 'CIDR')
+        {
```
Comment (anotherchrisberry) [inline app/scripts/modules/openstack/securityGroup/transformer.js]: move brace to line above

File: app/scripts/modules/openstack/securityGroup/configure/wizard/rules.controller.js [hunk 5]
```diff
@@ -4,13 +4,101 @@ let angular = require('angular');
...
+
+    $scope.remoteSecurityGroupSelected = function(indx, remoteSecurityGroupId) {
+      var rule = $scope.securityGroup.rules[indx];
+      if(remoteSecurityGroupId === 'CIDR') {
+        if(rule.prevcidr === '') {
+          rule.cidr = '0.0.0.0/0';
+          rule.prevcidr = '0.0.0.0/0';
+        }
+        else
+        {
```
Comment (anotherchrisberry) [inline app/scripts/modules/openstack/securityGroup/configure/wizard/rules.controller.js]: brace on wrong line

File: app/scripts/modules/openstack/securityGroup/configure/wizard/rules.controller.js [hunk 6]
```diff
@@ -4,13 +4,101 @@ let angular = require('angular');
...
+          rule.cidr = rule.prevcidr;
+        }
+      }
+      else
+      {
```
Comment (anotherchrisberry) [inline app/scripts/modules/openstack/securityGroup/configure/wizard/rules.controller.js]: move brace up

File: app/scripts/modules/openstack/common/validateType.directive.js [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+        {
```
Comment (anotherchrisberry) [inline app/scripts/modules/openstack/common/validateType.directive.js]: move brace up

File: app/scripts/modules/openstack/securityGroup/configure/wizard/rules.controller.js [hunk 8]
```diff
@@ -4,13 +4,101 @@ let angular = require('angular');
...
+       });
+    };
+
     this.removeRule = function(i) {
       $scope.securityGroup.rules.splice(i, 1);
     };
+
+    $scope.loadSecurityGroups = function(account, region) {
+      if(account !== undefined && region !== undefined)
+      {
```
Comment (anotherchrisberry) [inline app/scripts/modules/openstack/securityGroup/configure/wizard/rules.controller.js]: move brace

File: app/scripts/modules/openstack/securityGroup/details/details.html [hunk 9]
```diff
@@ -44,18 +44,36 @@ <h3 select-on-dbl-click>
         <dd>{{securityGroup.description}}</dd>
       </dl>
     </collapsible-section>
-    <collapsible-section heading="Rules ({{securityGroup.ipRangeRules.length || 0}})" expanded="{{!!(securityGroup.ipRangeRules && securityGroup.ipRangeRules.length)}}">
+    <collapsible-section heading="Ip Range Rules ({{securityGroup.ipRangeRules.length || 0}})" expanded="{{!!(securityGroup.ipRangeRules && securityGroup.ipRangeRules.length)}}">
```
Comment (anotherchrisberry) [inline app/scripts/modules/openstack/securityGroup/details/details.html]: Should probably be `IP`, not `Ip`

**Pull request merged**